File: V2/DIDx509/Models/X509Name.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace DIDx509.Models;

using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

/// <summary>
/// Represents an X.509 Name (Subject or Issuer) according to the DID:X509 JSON data model.
/// Maps X.509 distinguished name attributes to key-value pairs.
/// </summary>
public sealed class X509Name
{
    private readonly Dictionary<string, string> InternalAttributes;

    /// <summary>
    /// Gets the attributes dictionary (key is label or OID, value is UTF-8 string).
    /// </summary>
    public IReadOnlyDictionary<string, string> Attributes { get; }

    /// <summary>
    /// Initializes a new instance of the <see cref="X509Name"/> class.
    /// </summary>
    /// <param name="attributes">The name attributes.</param>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="attributes"/> is <see langword="null"/>.</exception>
    public X509Name(IDictionary<string, string> attributes)
    {
        InternalAttributes = new Dictionary<string, string>(attributes ?? throw new ArgumentNullException(nameof(attributes)), StringComparer.OrdinalIgnoreCase);
        Attributes = new ReadOnlyDictionary<string, string>(InternalAttributes);
    }

    /// <summary>
    /// Gets an attribute value by key (label or OID).
    /// </summary>
    /// <param name="key">The attribute key (label or OID).</param>
    /// <returns>The attribute value, or <see langword="null"/> if the attribute is not present.</returns>
    public string? GetAttribute(string key)
    {
        return InternalAttributes.TryGetValue(key, out var value) ? value : null;
    }

    /// <summary>
    /// Checks if the name contains all attributes from another name with matching values.
    /// </summary>
    /// <param name="other">The other name to compare against.</param>
    /// <returns><see langword="true"/> if all attributes and values in <paramref name="other"/> are present; otherwise, <see langword="false"/>.</returns>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="other"/> is <see langword="null"/>.</exception>
    public bool ContainsAll(X509Name other)
    {
        if (other == null)
        {
            throw new ArgumentNullException(nameof(other));
        }

        foreach (var kvp in other.Attributes)
        {
            if (!InternalAttributes.TryGetValue(kvp.Key, out var value) || !string.Equals(value, kvp.Value, StringComparison.Ordinal))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>
    /// Common Name (CN).
    /// </summary>
    public string? CN => GetAttribute(DidX509Constants.AttributeCN);

    /// <summary>
    /// Locality (L).
    /// </summary>
    public string? L => GetAttribute(DidX509Constants.AttributeL);

    /// <summary>
    /// State or Province (ST).
    /// </summary>
    public string? ST => GetAttribute(DidX509Constants.AttributeST);

    /// <summary>
    /// Organization (O).
    /// </summary>
    public string? O => GetAttribute(DidX509Constants.AttributeO);

    /// <summary>
    /// Organizational Unit (OU).
    /// </summary>
    public string? OU => GetAttribute(DidX509Constants.AttributeOU);

    /// <summary>
    /// Country (C).
    /// </summary>
    public string? C => GetAttribute(DidX509Constants.AttributeC);

    /// <summary>
    /// Street Address (STREET).
    /// </summary>
    public string? STREET => GetAttribute(DidX509Constants.AttributeSTREET);
}

File: native/rust/did/x509/src/models/x509_name.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/// An X.509 Distinguished Name attribute
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct X509NameAttribute {
    /// The attribute label (e.g., "CN", "O", "C")
    pub label: String,
    
    /// The attribute value
    pub value: String,
}

impl X509NameAttribute {
    /// Create a new X.509 name attribute
    pub fn new(label: String, value: String) -> Self {
        Self { label, value }
    }
}

/// An X.509 Distinguished Name (DN)
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct X509Name {
    /// The list of attributes in the DN
    pub attributes: Vec<X509NameAttribute>,
}

impl X509Name {
    /// Create a new X.509 name
    pub fn new(attributes: Vec<X509NameAttribute>) -> Self {
        Self { attributes }
    }

    /// Create an empty X.509 name
    pub fn empty() -> Self {
        Self {
            attributes: Vec::new(),
        }
    }

    /// Get the value of an attribute by label (case-insensitive)
    pub fn get_attribute(&self, label: &str) -> Option<&str> {
        self.attributes
            .iter()
            .find(|attr| attr.label.eq_ignore_ascii_case(label))
            .map(|attr| attr.value.as_str())
    }

    /// Get the Common Name (CN) attribute value
    pub fn common_name(&self) -> Option<&str> {
        self.get_attribute("CN")
    }

    /// Get the Organization (O) attribute value
    pub fn organization(&self) -> Option<&str> {
        self.get_attribute("O")
    }

    /// Get the Country (C) attribute value
    pub fn country(&self) -> Option<&str> {
        self.get_attribute("C")
    }
}
